### ISSUE TITLE
Obsoleate _enter_tree _exiting_tree and _exited_tree in favor for onready only scripts

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#2606]
### Describe the project you are working on
Basic asteroids clone

### Describe the problem or limitation you are having in your project
Its not trivial to make accessor properties due to the fact scripts are initialized and ran before the scene is ready

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
I suggest making the tree signals and hooks obsolete in favor in _ready and _destroying and _destroyed signals and hooks

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Basically scripts would not even be initialized until the scene is ready and the node with the script is ready.
This means properties would not be set until the ready state and on ready variables would become obsolete.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
It can be worked around but its not trivial for new users and people coming from other game engines.
This would fix that issue

### Is there a reason why this should be core and not an add-on in the asset library?
Just google how many `godot get_node returns null` topics and you will see that its not just me.

**Comment by zinnschlag** [comment]
> Basically scripts would not even be initialized until the scene is ready and the node with the script is ready.
This means properties would not be set until the ready state and on ready variables would become obsolete.

That would make it impossible to configure a node before adding it to the tree via code. Nope.

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/47904.

**Comment by Shadowblitz16** [comment]
@zinnschlag  you can configure the node with the constructor like all other languages or with initalizer syntax if gdscript ever supports it.

can you explain what you need to configure before the node is ready?

**Comment by zinnschlag** [comment]
@Shadowblitz16 No, if your proposal would be implemented you could not configure with the initialiser (even if gdscript supported it). The sequence is _init -> _ready. If you only initialise at _ready, there won't be an _init anymore. And you can't do a manual init function either. This completely kills the ability to build/prepare a scene in another thread, if there is any configuration required beyond just attaching node. You would have to create a separate data structure alongside it and then apply it after ready. Cumbersome and basically just reproducing the 2-stage init we currently have.

**Comment by LikeLakers2** [comment]
My problem with this proposal is that we'd needlessly be removing functionality. If `_enter_tree` were removed, those who are trying to set up variables with it are just going to transition to putting those `get_node` statements in `_ready`...

Thing is, using `_ready` is what you should already be doing. [The documentation for `Node._ready`](https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-method-ready) specifies that it runs only after a node and *all* of its children are added to the scene tree, which immediately solves the problem of `get_node` returning null when it shouldn't be.

Additionally, if a user wants to use `onready`, any who have read the documentation should already be aware that this:

```
onready var my_sprite = get_node("Sprite")
```

is equivalent to this:

```
var my_sprite

func _ready():
	my_sprite = get_node("Sprite")
```

I should also point out: [`Node`'s description](https://docs.godotengine.org/en/latest/classes/class_node.html#description) also includes a couple sections dedicated to pointing out what virtual calls get called when. If users are not reading that and assuming that `_enter_tree` is called when they've already been added to the scene, I think that's their fault.

That all said, if one were to be looking solely at the documentation for [`Node._enter_tree`](https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-method-enter-tree), it's not entirely clear that there is no access to children nodes in the scene tree, when that is called. So I think this could best be solved by updating the documentation.

**Comment by Shadowblitz16** [comment]
@LikeLakers2 
The issue is there is no such things as onready functions or export properties.

So when you try to link two node's properties together (forward them so they are editable after inheritance) it throws a error the the node doesn't exist.

this is because the script is ran before the scene is ready.

**Comment by LikeLakers2** [comment]
@Shadowblitz16 I'm not sure I understand. The code I put in my comment is perfectly valid code, last I checked. So would you care to elaborate on what you mean?

**Comment by hilfazer** [comment]
> Just google how many godot get_node returns null topics and you will see that its not just me.

There can be various sources of that problem, e.g. here
https://godotengine.org/qa/8666/why-is-my-get_node-returning-null
the user autoloaded a script instead of a scene.

And what would happen if someone tries to call a method on a scene that hasn't been added to the tree yet? An error saying "You can't call this node's methods yet. Please add it too the tree first."? That's also one of the reasons of the "get_node returns null" problems:
https://godotengine.org/qa/22321/dont-understand-why-get_node-is-returning-null

**Comment by Shadowblitz16** [comment]
@hilfazer you have a good point.

what would you guys think of onready exports, functions and classes? would that fix the issue?
I don't want to create another useless issue people are going to downvote and disregard.

**Comment by Shadowblitz16** [comment]
closing in favor of https://github.com/godotengine/godot-proposals/issues/2613
